woocommerce/google-listings-and-ads
# Aggregate intervals from free and paid campaigns, render programs report w/o waiting for secondary request

### Changes proposed in this Pull Request:

- Aggregate intervals from free and paid campaigns. (034b9f1)
	Fixes #589
- Refactor `useProgramsReport` hook, to reduce unnecessary computation. (681e6bd)
	Memoize aggregated totals and intervals, re-calculate only when actually needed.
	
	Separate `secondary` report calls, so `Chart`, `CompareTable` could render earlier.
	Render `SummarySection` with only the `primary` data while `secondary` is still loading.
	Implements part of #654
- Remove unused mocked programs data (f8ebeb8)


### Screenshots:

paid and free program intervals are aggregated to render chart:
![Screencast showing how paid and free program intervals are aggregated to render chart correctly](https://user-images.githubusercontent.com/17435/119563562-fdca1700-bda7-11eb-8e76-56df1f04883e.gif)
UI is being rendered before secondary response arrives:
![Screencast showing how UI is being rendered before secondary response arrives](https://user-images.githubusercontent.com/17435/119571652-e4c66380-bdb1-11eb-9aef-96eccd53669f.gif)

### Detailed test instructions:

0. Make sure you have some test data, or mock it with 
	  ```php
	  add_filter(
	      'gla_prepared_response_ads-reports-products',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/products.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_mc-reports-products',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/products.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_ads-reports-programs',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/ads/reports/programs.json' ), true ) ?: [];
	      },
	  );
	  add_filter(
	      'gla_prepared_response_mc-reports-programs',
	      function( $response ) {
	  	    return json_decode( file_get_contents( __DIR__ . '/tests/mocks/mc/reports/programs.json' ), true ) ?: [];
	      },
	  );
	  ```
1. Open Programs report page [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Freports%2Fprograms)
2. Switch between paid, free, and all programs and see that chart intervals add up.

---
Hard to observe without advanced mocking, as may be racy.
1. If the request for the previous period loads significantly longer than for the current one
2. You should see the entire root page be rendered, except deltas
3. Then update deltas eventually


### Things to consider:
- [ ] Add some tests for border cases and error cases, which are hard to catch using regular clicking without real data.
- [ ] Extend the enhancement to lazy load deltas to Products https://github.com/woocommerce/google-listings-and-ads/issues/654
- [ ] Add more explicit preloading of instead of `N/A` I'd also leave it to #654
- [ ] I hope this PR makes as closer not further away from https://github.com/woocommerce/google-listings-and-ads/issues/652

### Changelog Note:

> Fix programs report chart, aggregate intervals for free+paid programs.
> Improve UX by rendering programs report just after largest data chunk comes, without waiting for detlas.
